src/layer/unitary.rs

```diff
@@ -0,0 +1,465 @@
+            rec_t0 = to_complex(ltex.recurrences[0].clone());
+
+            // We normalize Householder parameters;
+            let sqrNorm = af::norm(&weight4, af::NormType::VECTOR_2, 1., 1.)as f32;
+            weight4 = af::div(&weight4, &sqrNorm, true);
+            ltex.weights[4] = to_real(weight4.clone());
+            
+            let sqrNorm = af::norm(&weight5, af::NormType::VECTOR_2, 1., 1.)as f32;
+            weight5 = af::div(&weight5, &sqrNorm, true);
+            ltex.weights[5] = to_real(weight5.clone());
+        }
+        
+        
+        
+      
+        // we compute h_t+1 = sigma1(W*h_t + V*x_t + b1) 
+        let wh = wh(weight1.clone()
+                    , weight4.clone()
+                    , ltex.optional[0].clone()
+                    , weight2.clone()
+                    , weight5.clone()
+                    , weight3.clone()
+                    , rec_t0.clone());
+
+        // In order to convert inputs.data into a complex array
+        let c_zeros = initializations::zeros::<Complex<f32>>(inputs.dims()); 
+
+        let c_inputs = af::add(inputs, &c_zeros, false);
+        let vx = af::matmul(&c_inputs
+                             , &weight0
+                             , MatProp::NONE
+                             , MatProp::NONE);
+
+        let new_h = af::add(&af::add(&wh, &vx, true)
+                            , &bias0
+                            , true);
+
+        let sigma_result = activations::get_activation(&ltex.activations[0]
+                                                       , &new_h).unwrap();
+
+
+        // we compute o_t = sigma2(U*h_t + b2)
+        let r_h = af::real(&sigma_result);
+        let c_h = af::imag(&sigma_result);
+        let concat_h = af::join(1, &r_h, &c_h);
+        let uh = af::matmul(&concat_h
+                            , &weight6
+                            , MatProp::NONE
+                            , MatProp::NONE);
+
+        let new_o = af::add(&uh, &bias1, true);
+       
+         
+        let out = activations::get_activation(&ltex.activations[1]
+                                              , &new_o).unwrap(); 
+        
+
+        if ltex.inputs.len() > t {
+            ltex.inputs[t] = c_inputs.clone();
+            ltex.recurrences[t+1] = to_real(sigma_result.clone());
+            ltex.outputs[t] = out.clone();
+        }
+        else{
+            ltex.inputs.push(c_inputs.clone());
+            ltex.recurrences.push(to_real(sigma_result.clone()));
+            ltex.outputs.push(out.clone());
+        }
+        //println!("{}", &(af::norm(&ltex.recurrences[t], af::NormType::VECTOR_2, 1.,1.)as f32));
+        ltex.current_unroll += 1;
+
+        (out.clone(), None)
+    }
+
+
+
+    fn backward(&self, params: Arc<Mutex<Params>>, delta: &Array) -> Array {
+        let mut ltex = params.lock().unwrap();
+        ltex.current_unroll -= 1;
+        let t = ltex.current_unroll;
+        let t_max = ltex.outputs.len()-1;
```

Comment:
This is probably not going to work. The data is passed in batch by batch from the sequential model. It isn't a 3d tensor when it gets here. If you need t_max I suggest placing it as a member in the params structure and using it from there.